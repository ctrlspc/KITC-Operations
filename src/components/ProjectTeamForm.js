import React from 'react';
import _ from 'lodash';
import Checkbox from 'rc-checkbox';

export default class ProjectTeamForm extends React.Component {
 
  constructor(props) {
    super(props)
    const selectedUsers = props.team || [];
    this.state = {
      error: false,
      selectedUsers
    };
  };

  onCheckBoxChange = (event) => {
const checkBox = event.target;
    if (checkBox.checked) {
      if (!_.includes(this.state.selectedUsers, checkBox.name )) {
        this.setState(() => ({selectedUsers: _.concat(this.state.selectedUsers, checkBox.name)}));
      }
    } else {
      const selectedUsers = this.state.selectedUsers.filter((user) => user !== checkBox.name)
      this.setState(() => ({selectedUsers}));
    }
  };

  onSubmit = (e) => {
    e.preventDefault();
    this.props.onSubmit(this.state.selectedUsers);
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p> }
        {
          this.props.users.map((user) => (
            <label className="label" key={user.uid}>
              <input
                className="checkbox-input"
                type="checkbox" 
                name={user.uid} 
                checked={_.includes(this.state.selectedUsers, user.uid)}
                onChange={this.onCheckBoxChange}/>
              {user.displayName}
            </label>
          ))
        }
        <div className="form__section">
          <button className="button">Save Team</button>
        </div>
      </form>
    );
  };
};