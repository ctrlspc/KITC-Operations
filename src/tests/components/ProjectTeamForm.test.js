import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';

import ProjectTeamForm from '../../components/ProjectTeamForm';
import profiles from '../fixtures/profiles';

test('should render a checkbox for all users', () => {
  const wrapper = shallow(<ProjectTeamForm users={profiles} team={[]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should select those users that are passed into the team prop', () => {
  const team = _.map([profiles[0], profiles[1]], (profile) => profile.uid);
  const wrapper = shallow(<ProjectTeamForm users={profiles} team={team}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should set a team member as selected in response to checkbox selection', () => {
  const wrapper = shallow(<ProjectTeamForm users={profiles}/>);
  wrapper.find(`input[name="${profiles[0].uid}"]`).at(0).simulate('change', {target: {checked:true, name:profiles[0].uid}});
  expect(wrapper.state('selectedUsers')).toEqual([profiles[0].uid]);
});

test('should unset a team member as selected in response to checkbox deselection', () => {
  const team = _.map([profiles[0], profiles[1]], (profile) => profile.uid);
  const wrapper = shallow(<ProjectTeamForm users={profiles} team={team}/>);
  expect(wrapper.state('selectedUsers')).toEqual([profiles[0].uid, profiles[1].uid]);
  wrapper.find(`input[name="${profiles[0].uid}"]`).at(0).simulate('change', {target: {checked:false,name:profiles[0].uid}});
  expect(wrapper.state('selectedUsers')).toEqual([profiles[1].uid]);
});

test('should handle onSubmit', () => {
  const onSubmitSpy = jest.fn();
  const preventDefaultSpy = jest.fn();
  const team = _.map([profiles[0], profiles[1]], (profile) => profile.uid);
  const wrapper = shallow(<ProjectTeamForm users={profiles} team={team} onSubmit={onSubmitSpy}/>);
  wrapper.find('form').simulate('submit',{preventDefault:preventDefaultSpy});
  expect(preventDefaultSpy).toHaveBeenCalled();
  expect(onSubmitSpy).toHaveBeenCalledWith([profiles[0].uid, profiles[1].uid]);
});