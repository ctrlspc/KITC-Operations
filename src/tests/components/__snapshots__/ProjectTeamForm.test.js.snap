// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a checkbox for all users 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <label
    className="label"
    key="123"
  >
    <input
      checked={false}
      className="checkbox-input"
      name="123"
      onChange={[Function]}
      type="checkbox"
    />
    Mr Test
  </label>
  <label
    className="label"
    key="223"
  >
    <input
      checked={false}
      className="checkbox-input"
      name="223"
      onChange={[Function]}
      type="checkbox"
    />
    Mr Testy
  </label>
  <label
    className="label"
    key="323"
  >
    <input
      checked={false}
      className="checkbox-input"
      name="323"
      onChange={[Function]}
      type="checkbox"
    />
    Mr Grumpy
  </label>
  <div
    className="form__section"
  >
    <button
      className="button"
    >
      Save Team
    </button>
  </div>
</form>
`;

exports[`should select those users that are passed into the team prop 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <label
    className="label"
    key="123"
  >
    <input
      checked={true}
      className="checkbox-input"
      name="123"
      onChange={[Function]}
      type="checkbox"
    />
    Mr Test
  </label>
  <label
    className="label"
    key="223"
  >
    <input
      checked={true}
      className="checkbox-input"
      name="223"
      onChange={[Function]}
      type="checkbox"
    />
    Mr Testy
  </label>
  <label
    className="label"
    key="323"
  >
    <input
      checked={false}
      className="checkbox-input"
      name="323"
      onChange={[Function]}
      type="checkbox"
    />
    Mr Grumpy
  </label>
  <div
    className="form__section"
  >
    <button
      className="button"
    >
      Save Team
    </button>
  </div>
</form>
`;
